src/backends/gpu/matrix_state_interleaved.hpp

```diff
@@ -0,0 +1,255 @@
+            auto src = cell_index_p[mtx];
+            auto dst = block*(block_dim()*padded_size) + lane;
+            for (auto i: make_span(0, len)) {
+                // the p indexes are always relative to the start of the p vector.
+                // the addition and subtraction of dst and src respectively is to convert from
+                // the original offset to the new padded and permuted offset.
+                p_tmp[dst+block_dim()*i] = dst + block_dim()*(p[src+i]-src);
+            }
+        }
+
+        d   = array(total_storage);
+        u   = array(total_storage);
+        rhs = array(total_storage);
+        parent_index = memory::make_const_view(p_tmp);
+
+        //
+        //  Calculate the invariant part of the matrix diagonal and the
+        //  upper diagonal on the host, then copy to the device.
+        //
+
+        std::vector<value_type> invariant_d_tmp(p.size(), 0);
+        std::vector<value_type> u_tmp(p.size(), 0);
+        auto face_conductance_tmp = memory::on_host(face_cond);
+        for(auto i: util::make_span(1u, p.size())) {
+            auto gij = face_conductance_tmp[i];
+
+            u_tmp[i] = -gij;
+            invariant_d_tmp[i] += gij;
+            invariant_d_tmp[p[i]] += gij;
+        }
+
+        // helper that converts to interleaved format on the host, then copies to device
+        // memory, for use as an rvalue in an assignemt to a device vector.
+        auto interleave = [&] (std::vector<T>const& x) {
+            return memory::on_gpu(
+                flat_to_interleaved(x, sizes_p, cell_index_p, block_dim(), num_mtx, padded_size));
+        };
+        u           = interleave(u_tmp);
+        invariant_d = interleave(invariant_d_tmp);
+        cv_capacitance = interleave(cv_cap);
+
+        matrix_sizes = memory::make_const_view(sizes_p);
+        matrix_index = memory::make_const_view(cell_index_p);
+
+        // allocate space for storing the un-interleaved solution
+        solution = array(p.size());
+    }
+
+    // the number of matrices stored in the matrix state
+    int num_matrices() const {
+        return matrix_sizes.size();
+    }
+
+    // the full padded matrix size
+    int padded_matrix_size() const {
+        return padded_size;
+    }
+
+    // Assemble the matrix
+    // Afterwards the diagonal and RHS will have been set given dt, voltage and current
+    //   dt      [ms]
+    //   voltage [mV]
+    //   current [nA]
+    void assemble(value_type dt, const_view voltage, const_view current) {
+        constexpr auto bd = impl::block_dim();
+        constexpr auto lw = impl::load_width();
+        constexpr auto block_dim = bd*lw;
+
+        // the number of threads is threads_per_matrix*num_mtx
+        const auto num_blocks = impl::block_count(num_matrices()*lw, block_dim);
+
+        assemble_matrix_interleaved<value_type, size_type, bd, lw, block_dim>
+            <<<num_blocks, block_dim>>>
+            ( d.data(), rhs.data(), invariant_d.data(),
```

Comment:
Code style: inconsistent spacing. (Space after opening parenthesis, no space before closing; elsewhere we have no space after opening parentheses.)